### ISSUE TITLE
implement learning approach for robot policy

### ISSUE THREAD

**Issue by mensch72** [mensch72/empo#49]
We're in the process of implementing both a backward-induction aproach and a model-based learning approach to approximate the quantities defined in equations (1)-(9) of the attached paper (please read it before working on this issue!). 

[empo-paper.pdf](https://github.com/user-attachments/files/24312033/empo-paper.pdf)

This has two phases: phase 1 computes equations (1)-(3), phase 2 computes equations (4)-(9). The backward induction code goes into `src/empo/backward_induction.py`, the learning-based one into `src/empo/nn_based`. We already have a first version for both approaches for phase 1 only. We also have incomplete code for the backward induction version of phase 2 (ignore that please). This issue is about implementing a first, vanilla version of the learning approach to phase 2 in the multigrid environment, using the same coding style as the existing code for phase 1. In particular, it is important throughout to clearly distinguish multigrid-specific code (like specific encoders) from environment-agnostic code, like we did for phase 1! 

Our approach will differ from the paper's equations only slightly, in the following way:

- While the paper discusses a single robot $r$, we consider several robots that are however coordinated by a single AI that uses a policy to choose all robots' actions at the same time and possibly correlated. In other words, the paper's $r$ denotes *the robot fleet* here and the paper's "robot action" $a_r$ denotes a combination of actions, one for each robot.
- We use expected values instead of sums in equations (7) and (8) to simplify the learning via stochastic approximation:
  - In equation (7), we use an expected value $E_{g_h\sim possible_goal_sampler(h)}$ instead of using $\sum_{g_h\in G_h}$, where `possible_goal_sampler` is of type `PossibleGoalSampler` (see `possible_goal.py`). We use the same sampler in equation (4).
  - In equation (8), we use an expected value $E_{h\sim Unif(human_agent_indices)}$ instead of using $\sum_h$.

The approach has some things in common with (multi-agent) reinforcement learning, but there are some crucial differences that are explained in `docs/API.md`.

The overall idea is to simultaneously learn the following neural networks:

- A `class NeuralRobotStateActionValue(RobotStateActionValue)` (approximating $Q_r$ according to equation (4)) that takes as input a model state $s$ and has an output node for each combination $a_r$ of robot actions (so if there are 3 robots with 8 actions each, the network has 8**3 = 512 output nodes) which outputs the logit of that action combination's probability. 
- A `class NeuralRobotPolicy(RobotPolicy)` (approximating $\pi_r$ according to equation (5)) with the same input/output shape as `NeuralRobotStateActionValue(RobotStateActionValue)` that takes as input a model state $s$ and has an output node for each combination $a_r$ of robot actions which outputs the logit of that action combination's probability. We will sometimes choose to have this "network" simply be a straightforward computation of $pi_r(s)$ from $Q_r(s)$ according to equation (5), but we will sometime instead choose this to really be a separate neural network that is trained on the targets given by equation (5). The code should support both versions seamlessly via config params.
- A `class NeuralHumanGoalAchievementAbility(HumanGoalAchievementAbility)` (approximating $V^e_h$ according to equation (6)) taking as input a state, a human, and a goal (exactly like the existing network `MultiGridQNetwork`, whose method `encode_and_forward` takes as inputs a `world_model` and `state` ($s$), a `query_agent_idx` ($h$), and a `goal` ($g_h$)), and outputting a single number between 0 and 1 (using a `SoftClamp` with `feasible_range` (0,1)). 
- A `class NeuralAggregateGoalAchievementAbility(AggregateGoalAchievementAbility)` (approximating $X_h$ according to equation (7)) taking as input a state and a human, trained with a loss function suitable for learning expected values of random variables bounded between 0 and 1.
- A `class NeuralIntrinsicRobotReward(IntrinsicRobotReward)` (approximating $U_r$) taking as input a state. Its last hidden layer approximates the expected value $y := E_{h\sim Unif} X_h(s)^{-\xi}$, and the output neuron then outputs $U_r(s) = -y^{\eta}$, see equation (8). This is trained with a loss function for $y$ suitable for learning expected values of random variables bounded from below by $1$ and having a heavy right tail (because $X_h(s)^{-\xi}$ is in $(1,\infty)$ since $X_h(s)\in [0,1]$ and $\xi\ge 1$).
- A `class NeuralRobotStateValue(RobotStateValue)` (approximating $V_r$ according to equation (9)) that takes as input a model state $s$ and outputs a single number. This should use a standard loss function suitable for value networks in RL problems, maybe configurable.

These networks are trained on training data points $(s, a_r, g, a_H, s')$ generated from rollouts that use:
- The current `neural_robot_policy` or optionally a frozen version of it, as in Double Q Learning. Here we might add some additional `epsilon_r`-greedy style exploration as described in `docs/API.md` that we phase out according to some schedule (while we do *not* make $\beta_r$ or $\beta_h$ go to infinity as they are part of the robot's beliefs or normative choices!).
- A combination of goals $g=(g_h)_h$, one for each human, generated randomly at the beginning of each episode and updated randomly whenever some human has achieved their goal, using a given `PossibleGoalSampler`.
- A fixed given `HumanPolicyPrior` (representing $\pi_H$, either handcoded or learned in phase 1 before).
- A successor state $s'$ generated by `world_model.step()`.

While all networks are trained on the same rollouts, the training hyperparameters for each networks should be independently configurable: learning rates, batch sizes, frozen network update intervals, etc. 

Whenever some of the networks occurs in the RHS of the definition of the training target for another network (like $Q_r$ occurs in the RHS for $V_r$),, we should be able to choose to use either the current network or an earlier frozen version (like in Double DQN). 

To complete the issue, also provide some tests and example scripts similar to the ones for phase 1. In the example script, train on the same gridworld ensemble as in `examples/random_multigrid_ensemble_demo.py` and use the `HeuristicPotentialPolicy` from `empo.human_policy`.
